src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,269 @@
+namespace AlpnProxy {
+namespace {
+
+std::unique_ptr<::Envoy::Buffer::OwnedImpl> constructProxyHeaderData(
+    const Envoy::ProtobufWkt::Any& metadata) {
+  envoy::tcp::alpnproxy::header::AlpnProxyHeaderProto proxy_header;
+  proxy_header.mutable_metadata()->CopyFrom(metadata);
+
+  AlpnProxyInitialHeader initial_header;
+  std::string proxy_header_str = proxy_header.SerializeAsString();
+  // Converting from host to network byte order so that most significant byte is
+  // placed first.
+  initial_header.magic = absl::ghtonl(AlpnProxyInitialHeader::magic_number);
+  initial_header.data_size = absl::ghtonl(proxy_header_str.length());
+
+  ::Envoy::Buffer::OwnedImpl initial_header_buffer{
+      absl::string_view(reinterpret_cast<const char*>(&initial_header),
+                        sizeof(AlpnProxyInitialHeader))};
+  auto proxy_header_buffer =
+      std::make_unique<::Envoy::Buffer::OwnedImpl>(proxy_header_str);
+  proxy_header_buffer->prepend(initial_header_buffer);
+  return proxy_header_buffer;
+}
+
+std::unique_ptr<Envoy::ProtobufWkt::Struct> getStructMetadataFromAny(
+    const Envoy::ProtobufWkt::Any& any_metadata) {
+  static const std::string& struct_type =
+      Envoy::ProtobufWkt::Struct::default_instance()
+          .GetDescriptor()
+          ->full_name();
+  absl::string_view type = any_metadata.type_url();
+  size_t pos = type.find_last_of('/');
+  if (pos != absl::string_view::npos) {
+    type = type.substr(pos + 1);
+  }
+  Envoy::ProtobufWkt::Struct struct_metadata;
+  if (type == struct_type) {
+    struct_metadata.MergeFromString(any_metadata.value());
+  } else {
+    auto struct_value_map = struct_metadata.mutable_fields();
+    google::protobuf::Value& map_value =
+        (*struct_value_map)[any_metadata.type_url()];
+    (*map_value.mutable_string_value()) = any_metadata.SerializeAsString();
+  }
+  return std::make_unique<Envoy::ProtobufWkt::Struct>(struct_metadata);
+}
+
+}  // namespace
+
+AlpnProxyConfig::AlpnProxyConfig(const std::string& stat_prefix,
+                                 const std::string& protocol,
+                                 const std::string& node_metadata_id,
+                                 const FilterDirection& filter_direction,
+                                 Stats::Scope& scope)
+    : scope_(scope),
+      stat_prefix_(stat_prefix),
+      protocol_(protocol),
+      node_metadata_id_(node_metadata_id),
+      filter_direction_(filter_direction),
+      stats_(generateStats(stat_prefix, scope)) {}
+
+Network::FilterStatus AlpnProxyFilter::onData(Buffer::Instance& data, bool) {
+  // No work needed if we have read connection protocol and its not the expected
+  // alpn protocol or if alpn protocol is found but we are done reading data.
+  if ((connection_protocol_read_ && !alpn_protocol_found_) ||
+      (alpn_protocol_found_ && read_done_)) {
+    return Network::FilterStatus::Continue;
+  }
+  if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
```

Comment:
Thanks for looking.. I think what was needed was !alpn_protocol_found...

Please check logic below:

If ALPN protocol is not found then in first time itself, connection_protocol_read_ will be true and will return from that if and alpn_protocol_found will be false. Then in first if of the function it will return because of (connection_protocol_read_ && !alpn_protocol_found_ ) becoming true and it will not read protocol again and again.
if ALPN protocol found is true.. then first time it will read the protocol, not set connection_protocol_read but will set alpn_protocol_found to true. If we have more data to read, then later times, alpn_protocol_found will be true and it will not enter this if (!alpn_protocol_found_ &&
      read_callbacks_->connection().nextProtocol() != config_->protocol_) and not read protocol again and again.

Reply:
I am not sure. Maybe it would be better to use a state machine instead of the bits and just draw a simple chart for the transition diagram?

Reply:
Done.. extended the state machine I had for reading headers to reading protocol too.